src/trace/log.zig

```diff
@@ -1,366 +1,463 @@
         self.* = .{
-            .allocator = allocator,
-            .max_level = max_level,
+            .allocator = config.allocator,
+            .log_allocator = recycle_fba.allocator(),
+            .log_allocator_state = recycle_fba,
+            .max_buffer = max_buffer,
             .exit_sig = AtomicBool.init(false),
+            .max_level = config.max_level,
             .handle = null,
-            .channel = Channel(*StandardEntry).create(allocator) catch
-                @panic("could not allocate StandardEntry channel"),
+            .channel = Channel(logfmt.LogMsg).create(config.allocator) catch
+                @panic("could not allocate LogMsg channel"),
         };
+        self.handle = try std.Thread.spawn(.{}, run, .{self});
         return self;
     }
 
-    pub fn spawn(self: *Self) void {
-        self.handle = std.Thread.spawn(.{}, StandardErrLogger.run, .{self}) catch @panic("could not spawn Logger");
-    }
-
     pub fn deinit(self: *Self) void {
-        if (self.handle) |handle| {
+        if (self.handle) |*handle| {
+            std.time.sleep(std.time.ns_per_ms * 5);
             self.exit_sig.store(true, .seq_cst);
             handle.join();
         }
         self.channel.deinit();
+        self.log_allocator_state.deinit();
         self.allocator.destroy(self.channel);
+        self.allocator.destroy(self.log_allocator_state);
         self.allocator.destroy(self);
     }
 
-    fn run(self: *Self) void {
-        const sink = StdErrSink{};
-
-        while (!self.exit_sig.load(.seq_cst)) {
-            std.time.sleep(std.time.ns_per_ms * 5);
-
-            while (self.channel.receive()) |e| {
-                sink.consumeEntry(e);
-                e.deinit();
-            }
-        }
-    }
-
-    pub fn field(self: *Self, name: []const u8, value: anytype) Entry {
-        var e = Entry.init(self.allocator, self.channel, self.max_level);
-        return e.field(name, value);
+    pub fn logger(self: *Self) Logger {
+        return .{ .standard = self };
     }
 
-    pub fn info(self: *Self, msg: []const u8) void {
-        self.log(.info, msg);
+    pub fn scopedLogger(self: *Self, comptime new_scope: anytype) ScopedLogger(new_scope) {
+        return .{ .standard = self };
     }
 
-    pub fn debug(self: *Self, msg: []const u8) void {
-        self.log(.debug, msg);
-    }
-
-    pub fn warn(self: *Self, msg: []const u8) void {
-        self.log(.warn, msg);
-    }
-
-    pub fn err(self: *Self, msg: []const u8) void {
-        self.log(.err, msg);
+    pub fn run(self: *Self) void {
+        while (!self.exit_sig.load(.acquire)) {
+            while (self.channel.receive()) |message| {
+                const writer = std.io.getStdErr().writer();
+                { // Scope to limit the span of the lock on std err.
+                    std.debug.lockStdErr();
+                    defer std.debug.unlockStdErr();
+                    logfmt.writeLog(writer, message) catch {};
+                }
```

Comment:
```suggestion
                { 
                    const writer = std.io.getStdErr().writer();
                    std.debug.lockStdErr();
                    defer std.debug.unlockStdErr();
                    
                    logfmt.writeLog(writer, message) catch {};
                }
```

Reply:
This was based of the suggestion in the comment here https://github.com/Syndica/sig/pull/249#discussion_r1751986823

Reply:
this doesnt add lock contention or change the logic - it moves the writer variable within the block of the lock so its more clear where the `writer` variable is used

Reply:
[4e0edfa](https://github.com/Syndica/sig/pull/277/commits/4e0edfa4761a68e7d9c11beeb61796f54a83f64a)